### PR TITLE
[TF-7941] Add support for Terraform Test

### PR DESCRIPTION
**Pull request by aaabdelgany**
## Description

This PR adds support for the new Terraform Test Runs API.

We can create new kinds of registry modules (branch-based), upload configuration versions against registry modules, and create, monitor, and cancel test runs.

## Testing plan

I've ran numerous e2e tests manually using my local build of Terraform Core. You can see the PR in the Terraform repo that uses this to integrate the testing framework with TFC: https://github.com/hashicorp/terraform/pull/33710

I've added integration tests for the new API calls in this PR as well, but they are unfortunately untestable at the moment as the test endpoints are protected by a feature flag while the testing framework is going through the Beta rollout. This means that organisations created newly within the testing framework don't have access to the test runs endpoint.

### Testing output

I have executed the tests locally, by hacking the create organisation helper function to load an already whitelisted organisation from the staging environment instead of creating a new one and they've passed.

```
~/hashicorp/go-tfe (nexus/terraform_test_cli) > envchain go-tfe go test -run TestTestRuns -v ./...
?       github.com/hashicorp/go-tfe/examples/configuration_versions     [no test files]
?       github.com/hashicorp/go-tfe/examples/organizations      [no test files]
?       github.com/hashicorp/go-tfe/examples/registry_modules   [no test files]
?       github.com/hashicorp/go-tfe/examples/state_versions     [no test files]
?       github.com/hashicorp/go-tfe/examples/users      [no test files]
?       github.com/hashicorp/go-tfe/examples/workspaces [no test files]
?       github.com/hashicorp/go-tfe/mocks       [no test files]
=== RUN   TestTestRunsList
=== RUN   TestTestRunsList/without_list_options
=== RUN   TestTestRunsList/empty_list_options
=== RUN   TestTestRunsList/with_page_size
--- PASS: TestTestRunsList (5.82s)
    --- PASS: TestTestRunsList/without_list_options (0.23s)
    --- PASS: TestTestRunsList/empty_list_options (0.23s)
    --- PASS: TestTestRunsList/with_page_size (0.21s)
=== RUN   TestTestRunsRead
=== RUN   TestTestRunsRead/when_the_test_run_exists
=== RUN   TestTestRunsRead/when_the_test_run_does_not_exist
--- PASS: TestTestRunsRead (4.55s)
    --- PASS: TestTestRunsRead/when_the_test_run_exists (0.23s)
    --- PASS: TestTestRunsRead/when_the_test_run_does_not_exist (0.20s)
=== RUN   TestTestRunsCreate
=== RUN   TestTestRunsCreate/with_a_configuration_version
=== RUN   TestTestRunsCreate/without_a_configuration_version
=== RUN   TestTestRunsCreate/without_a_module
--- PASS: TestTestRunsCreate (4.09s)
    --- PASS: TestTestRunsCreate/with_a_configuration_version (0.29s)
    --- PASS: TestTestRunsCreate/without_a_configuration_version (0.00s)
    --- PASS: TestTestRunsCreate/without_a_module (0.00s)
=== RUN   TestTestRunsLogs
=== RUN   TestTestRunsLogs/when_the_log_exists
=== RUN   TestTestRunsLogs/when_the_log_does_not_exist
--- PASS: TestTestRunsLogs (11.90s)
    --- PASS: TestTestRunsLogs/when_the_log_exists (7.54s)
    --- PASS: TestTestRunsLogs/when_the_log_does_not_exist (0.21s)
=== RUN   TestTestRunsCancel
=== RUN   TestTestRunsCancel/when_the_run_exists
=== RUN   TestTestRunsCancel/when_the_run_does_not_exist
--- PASS: TestTestRunsCancel (4.63s)
    --- PASS: TestTestRunsCancel/when_the_run_exists (0.24s)
    --- PASS: TestTestRunsCancel/when_the_run_does_not_exist (0.21s)
PASS
ok      github.com/hashicorp/go-tfe     31.696s
```

### CODE REVIEW

File: test_run.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+}
+
+// Create a new test run with the given options.
+func (s *testRuns) Create(ctx context.Context, options TestRunCreateOptions) (*TestRun, error) {
+	if err := options.valid(); err != nil {
+		return nil, err
+	}
+
+	moduleID := RegistryModuleID{
+		Organization: options.RegistryModule.Organization.Name,
```
Comment (brandonc) [inline test_run.go:196]: Organization _could_ be nil here, causing a panic. You could check for this in TestRunCreateOptions valid

File: errors.go [hunk 2]
```diff
@@ -354,6 +354,10 @@ var (
 
 	ErrRequiredRegistryModule = errors.New("registry module is required")
 
+	ErrInvalidRegistryModuleID = errors.New("invalid value for registry module ID")
```
Comment (brandonc) [inline errors.go:357]: I don't think this is being used

**Review by brandonc — COMMENTED** [review]
Looks highly idiomatic! Just a few minor concerns

File: test_run.go (hunk 1 above)
Comment (liamcervante) [reply test_run.go:196]: Done!

File: errors.go (hunk 2 above)
Comment (liamcervante) [reply errors.go]: Removed!

**Pull request merged**